### DISCUSSION TITLE
3.1.60  problem，undefined symbol

### DISCUSSION THREAD

**Discussion by WenAce** [emscripten-core/emscripten#22066] [Q&A] [answered]
wasm-ld: error: C:\Users\xxx\AppData\Local\Temp\tmp55p8damxlibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_fullscreenchange_callback_on_thread，this is what I built my project and it shows the error，what is the problem，I use -pthread，-sPROXY_TO_PTHREAD=1 -sOFFSCREENCANVAS_SUPPORT=1 to build

**Comment by sbc100** [comment]
Is this a regression?  (i.e. did it work in previous version?)

Those symbols (`_emscripten_run_callback_on_thread` and `emscripten_set_fullscreenchange_callback_on_thread`) are part of libhtml5 (one in the JS html5 library and one in the native html5 library).

Are you share the full link command you are using?

**Reply by WenAce** [reply]
it did work on the 3.1.46 version, my full command   
            -WASM64=1
            -USE_WEBGL2=1
            -FULL_ES3=1
            -ENVIRONMENT=web,worker
            -FORCE_FILESYSTEM=1
            -JS_MATH=1
            -STACK_SIZE=100MB
            -EXIT_RUNTIME=1
            -USE_OFFSET_CONVERTER=1
            -NO_DISABLE_EXCEPTION_CATCHING=1
            -MODULARIZE=1								
            -EXPORT_ES6=1								
            -lembind  
            -pthread 
 -sPTHREAD_POOL_SIZE=16
 -sSHARED_MEMORY=1
             -mbulk-memory
             -Wl
             --shared-memory
             -WASM_WORKERS=1
            -PROXY_TO_PTHREAD=1
            -OFFSCREENCANVAS_SUPPORT=1
I have found that in the previous version(3.1.46) when I use the function emscripten_set_fullscreenchange_callback_on_thread, it is part of library_html5.js and the function is JSEvents.queueEventHandlerOnThread_iiii, but now is changed, it is __emscripten_run_callback_on_thread, The program encountered an error at the end of the compilation,

**Reply by sbc100** [reply]
Something looks a little off with that command line (a lot of `-s` seem to be missing.. perhaps just a copy and paste error of some kind?).  For example `-WASM64=1` doesn't work since its missing the `-s` at the beginning.. also the setting is called `-sMEMORY64` I think. 

BTW if you use `-pthread` you don't also need `-sSHARED_MEMORY` and you don't want to be passing direct linker flags such as `-Wl,--shared-memory` (emscripten takes care of that).

**Reply by WenAce** [reply]
in my project，I already use -s at the beginning，so it's fine,but it just show me some error.
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_webgl_create_context
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_fullscreenchange_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_mouseup_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_keydown_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_blur_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_keyup_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_set_offscreencanvas_size_on_thread. Required by emscripten_request_fullscreen_strategy
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_request_fullscreen_strategy
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_mousedown_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_mousemove_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_wheel_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_touchend_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_touchmove_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_touchcancel_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_set_offscreencanvas_size_on_thread. Required by emscripten_set_canvas_element_size
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_canvas_element_size
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_touchstart_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_resize_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_focus_callback_on_thread
wasm-ld: error: C:\Users\AppData\Local\Temp\tmpg1qg72pklibemscripten_js_symbols.so: undefined symbol: _emscripten_run_callback_on_thread. Required by emscripten_set_keypress_callback_on_thread

**Reply by sbc100** [reply]
Can you share the full error that shows all the wasm-ld command line flags.  That `_emscripten_run_callback_on_thread` symbol lives in libhtml5, but it seems like perhaps `-lhtml5` is not being passed to wasm-ld?  Can you confirm?

**Reply by WenAce** [reply]
we found the answer， which is the “-s strict=1”, it seems like it will make us link lib by ourselves, so now we don't use that， and we succeeded to build， thanks

**Reply by sbc100** [reply, marked as answer]
Yes, `-sSTRICT` implies `-sAUTO_NATIVE_LIBRARIES` which means you need to explicitly add `-lhtml5`.. or you could set `-sAUTO_NATIVE_LIBRARIES=1` explicitly.
